distributed/worker.py

```diff
@@ -3388,11 +3333,8 @@ def done_event():
             for d in self.in_flight_workers.pop(worker):
                 ts = self.tasks[d]
                 ts.done = True
-                if d in cancelled_keys:
-                    if ts.state == "cancelled":
-                        recommendations[ts] = "released"
-                    else:
-                        recommendations[ts] = "fetch"
+                if ts.state == "cancelled":
```

Comment:
I believe we should remove this special treatment. The bigger point of the transition system was to simplify these kind of clauses and allow us to make a recommendation without investigating start states. This did not work well all the time but in this case it works flawlessly and reduces complexity as intended.


I like the original transition log also better because a successful fetch should recommend a transition to memory. However, the state machine decides to forget instead. This is much more in line with how I would envision this system to work.

```diff
diff --git a/distributed/tests/test_cancelled_state.py b/distributed/tests/test_cancelled_state.py
index cab21a5c..74a039b7 100644
--- a/distributed/tests/test_cancelled_state.py
+++ b/distributed/tests/test_cancelled_state.py
@@ -322,10 +322,7 @@ async def test_in_flight_lost_after_resumed(c, s, b):
             ("free-keys", (fut1.key,)),
             (fut1.key, "resumed", "released", "cancelled", {}),
             # After gather_dep receives the data, the task is forgotten
-            ("receive-dep", a.address, {fut1.key}),
-            (fut1.key, "release-key"),
-            (fut1.key, "cancelled", "released", "released", {fut1.key: "forgotten"}),
-            (fut1.key, "released", "forgotten", "forgotten", {}),
+            (fut1.key, "cancelled", "memory", "released", {fut1.key: "forgotten"}),
         ],
     )

diff --git a/distributed/worker.py b/distributed/worker.py
index cc2ea229..3f6319fa 100644
--- a/distributed/worker.py
+++ b/distributed/worker.py
@@ -3333,9 +3333,7 @@ class Worker(ServerNode):
             for d in self.in_flight_workers.pop(worker):
                 ts = self.tasks[d]
                 ts.done = True
-                if ts.state == "cancelled":
-                    recommendations[ts] = "released"
-                elif d in data:
+                if d in data:
                     recommendations[ts] = ("memory", data[d])
                 elif busy:
                     recommendations[ts] = "fetch"
```

Reply:
Very happy to apply the patch if it doesn't deadlock elsewhere :stuck_out_tongue:

Reply:
To clarify: if a asks b for x, but either x responds that it doesn't have a replica or doesn't respond at all, this will trigger a cancelled->fetch transition. Is this the desired behaviour?

Reply:
Desired behavior is probably a bit much. It will do the right thing because we'll have

ts.done = True
cancelled -> fetch

which will recommend a cancelled->release so we're good.

I fully admit that the `ts.done` attribute in this case is very awkward. It basically encodes that this fetch transition originates from either the gather_dep result or from the execute result. Therefore, we could just as well remove the `ts.done` attribute and deal with these transition in the gather_dep/execute result the way you are proposing in this PR. When I introduced this (many months ago) I felt this would reduce code complexity (as in having fewer conditionals).
Given that we still have the `ts.done` attribute, I believe the patch I am proposing is the more idiomatic way but I'm happy to revisit this in a later iteration.

Reply:
I would also be in favor of removing `ts.done` eventually and having the logic in `gather_dep` and `execute` like here. Or maybe it's just a naming issue—`ts.done` is a pretty generic/ambiguous term. But I think from a https://github.com/dask/distributed/issues/5736 perspective, having this extra piece of state (`done`) that affects the behavior of transitions makes things harder to reason about. Though I do appreciate that it protects you from forgetting about these edge cases and having to check whether `ts.state == "cancelled"`.

Maybe this is over the top, but what if `done` was a state? Call it `fetched` and `executed`, since they might need different logic and I don't like overlapping the states of execution vs fetching anyway. Then you'd have different transition handlers for flight->fetched vs cancelled->fetched. Forgetting to handle the cancelled possibility would be an impossible transition error, instead of a bug and maybe deadlock.

Reply:
I don't think "fetched" or "executed" is a good idea - I'd rather look into moving away from intermediate states, not adding more.